woocommerce/woocommerce-ios
# Load/validate domain contact info before/after the form

Part of #8558 


## Description


When checking out a cart with a domain using domain credit, the contact info needs to be validated first. Otherwise, contact info validation errors are thrown. In this PR, two API requests are made before & after the domain contact info form:
- Load pre-existing contact info after selecting a domain and before entering the form. If it fails, then the contact info form is blank
- Validate contact info when tapping `Done`: if there are validation errors, a list of error messages is returned in a new error `DomainContactInfoError`

Since we're validating most of the fields on the server side, we want to always enable the `Done` CTA to validate the current form. However, the current address form has the `Done` CTA disabled by default. In order to always enable the `Done` CTA for this use case, I added a new parameter `isDoneButtonAlwaysEnabled: Bool` to `AddressFormViewModel` with a default value `false` to not affect the order editing/creation use case.

One note on the phone number field: it has to be in the format of a valid country code followed by a dot (for example +1.1233456). The current form doesn't support dot with the phone pad keyboard type, in a future subtask I'm going to add a new field for phone country code to improve the UX.

## Testing instructions


Prerequisite: a WC store with an annual WPCOM plan whose domain credit hasn't been claimed, you can create a site like this in Calypso with store credit

- Log in if needed, and continue with a WPCOM store
- Go to the Menu tab
- Tap on the settings CTA
- Tap the `Domain` row --> there should be a section about the free domain for the first year
- Tap `Claim Domain` --> domain selector with paid domains should be shown
- Select a domain and tap `Continue` --> the CTA should become a spinner while loading contact info, then the domain contact info form should be shown with the existing contact info pre-filled
- Remove the email field and tap `Done` --> an "invalid email" notice should be shown
- Enter a valid email but remove the first or last name, then tap `Done` --> an error notice should be shown
- Update the form with valid info (note the phone value needs to follow the format mentioned above) and tap `Done` --> the spinner should show up, then continue to the success screen
- Tap `Continue` to go back to domain settings --> the new domain should show up shortly

## Screenshots


local email validation error | phone validation error
-- | --
![Simulator Screen Shot - iPhone 14 - 2023-02-02 at 10 27 00](https://user-images.githubusercontent.com/1945542/216216244-64774876-fe6c-4317-9503-f6f3649fc9f4.png) | ![Simulator Screen Shot - iPhone 14 - 2023-02-02 at 10 27 17](https://user-images.githubusercontent.com/1945542/216216259-a7a2dfc3-5a29-4292-b9ac-09d9aa62519f.png)


---
